modules/mod-deep-learning/DeepLearningEffect.cpp

```diff
@@ -0,0 +1,182 @@
+/**********************************************************************
+
+   Audacity: A Digital Audio Editor
+   Audacity(R) is copyright (c) 1999-2021 Audacity Team.
+
+   DeepLearningEffect.cpp
+   Hugo Flores Garcia
+
+******************************************************************/
+/**
+
+*/
+/*******************************************************************/
+
+#include "DeepLearningEffect.h"
+
+#include <wx/stattext.h>
+
+#include "FileNames.h"
+#include "Shuttle.h"
+#include "ShuttleGui.h"
+#include "WaveTrack.h"
+
+#include "effects/LoadEffects.h"
+
+#include <torch/script.h>
+
+// DeepLearningEffectBase implementation
+
+std::string DeepLearningEffect::GetDeepEffectID()
+{ return "waveform-to-waveform";}
+
+const ComponentInterfaceSymbol DeepLearningEffect::Symbol
+{ XO("Deep Learning Effect") };
+
+// register source separation
+namespace{ BuiltinEffectsModule::Registration< DeepLearningEffect > reg; }
+
+DeepLearningEffect::DeepLearningEffect()
+{
+   SetLinearEffectFlag(false);
+}
+
+DeepLearningEffect::~DeepLearningEffect()
+{
+}
+
+// ComponentInterface implementation
+
+ComponentInterfaceSymbol DeepLearningEffect::GetSymbol()
+{
+   return Symbol;
+}
+
+TranslatableString DeepLearningEffect::GetDescription()
+{
+   return XO("The goal of audio source separation is to isolate "
+             "the sound sources in a given mixture of sounds.");
+}
+
+ManualPageID DeepLearningEffect::ManualPage()
+{
+   return L"Source_Separation"; 
+}
+
+// EffectDefinitionInterface implementation
+
+EffectType DeepLearningEffect::GetType()
+{
+   return EffectTypeProcess;
+}
+
+// Effect implementation
+
+// ProcessOne() takes a track, transforms it to bunch of buffer-blocks,
+// performs a forward pass through the deep model, and writes 
+// the output to new tracks. 
+bool DeepLearningEffect::ProcessOne(WaveTrack *leader,
+                           double tStart, double tEnd)
+{
+   if (leader == NULL)
+      return false;
+
+   // keep track of the sample format and rate,
+   // we want to convert all output tracks to this
+   sampleFormat origFmt = leader->GetSampleFormat();
+   int origRate = leader->GetRate();
+
+   // initialize source tracks, one for each source that we will separate
+   std::vector<WaveTrack::Holder> sourceTracks;
+   std::vector<std::string> sourceLabels = mModel->GetCard()->labels();
+   sourceTracks = CreateSourceTracks(leader, sourceLabels);
+   
+   // Initiate processing buffer, most likely shorter than
+   // the length of the selection being processed.
+   Floats buffer{ leader->GetMaxBlockSize() };
+
+   // get each of the blocks we will process
+   std::vector<BlockIndex> indices = GetBlockIndices(leader, tStart, tEnd);
+   for (BlockIndex block : indices)
+   {
```

Comment:
So, I understand that this effect is doing an outer loop over intervals of time, while any loops over channels are inner loops.

GetBlockIndices() looks like it was a lot of work in pursuit of efficiencies, but was there measurement of the benefit?

I think you are assuming too that the block boundaries for efficient access occur at the same times in the two channels, but that might be untrue in unusual cases.

The division into blocks should have no effect on the outputs of the model.  Can I be sure that is so?  Does the model know how to carry state from one pass of the loop to the next?  Is that exactly one of the points of the TensorFlow library, to process large amounts of data piecewise?

Reply:
If this pattern of efficient iteration over the blocks of a track is really important, I should think about how to extract it into a reusable utility, and also treat the multiple channel case, for reuse in other effects and analyzers.

Reply:
Is it possible to just process the track clip-by-clip? Would it cause any memory problems if the clip is too large? Using `track->GetBestBlockSize` was just the way a few other effects are implemented in Audacity, so I figured I would try to follow the pattern.